src/components/app.tsx

```diff
@@ -17,66 +19,78 @@ interface IState {
+    }
+  }
+  else {
+    db.connect({appMode, stores})
+      .then(() => {
+        if (appMode === "qa") {
+          const {qaClear, qaGroup} = urlParams;
+          if (qaClear) {
+            const cleared = (err?: string) => {
+              if (onQAClear) {
+                onQAClear(!err, err);
+              }
+            };
+            db.clear(qaClear)
+              .then(() => cleared())
+              .catch(cleared);
+          }
+          else if (qaGroup) {
+            db.leaveGroup().then(() => db.joinGroup(qaGroup));
+          }
+        }
+      })
+      .catch(ui.setError);
+  }
+}
+
 export const authAndConnect = (stores: IStores, onQAClear?: (result: boolean, err?: string) => void) => {
-  const {appMode, user, db, ui} = stores;
+  const {appMode, user, ui} = stores;
 
   authenticate(appMode, urlParams)
-    .then(({authenticatedUser, classInfo, problemId}) => {
+    .then(({authenticatedUser, classInfo, problemId, unitCode}) => {
       user.setAuthenticatedUser(authenticatedUser);
       if (classInfo) {
         stores.class.updateFromPortal(classInfo);
       }
       if (problemId) {
-        updateProblem(stores, problemId);
-      }
-
-      if (typeof (window as any).Rollbar !== "undefined") {
-        const _Rollbar = (window as any).Rollbar;
-        if (_Rollbar.configure) {
-          const config = { payload: {
-                  class: authenticatedUser.classHash,
-                  offering: authenticatedUser.offeringId,
-                  person: { id: authenticatedUser.id },
-                  problemId: problemId || "",
-                  problem: stores.problem.title,
-                  role: authenticatedUser.type,
-                  unit: stores.unit.title,
-                  version: stores.appVersion
-                }};
-          _Rollbar.configure(config);
-        }
-      }
-
-      if (appMode === "authed")  {
-        const { rawFirebaseJWT } = authenticatedUser;
-        if (rawFirebaseJWT) {
-          db.connect({appMode, stores, rawFirebaseJWT}).catch(ui.setError);
-        }
-        else {
-          ui.setError("No firebase token available to connect to db!");
-        }
-      }
-      else {
-        db.connect({appMode, stores})
-          .then(() => {
-            if (appMode === "qa") {
-              const {qaClear, qaGroup} = urlParams;
-              if (qaClear) {
-                const cleared = (err?: string) => {
-                  if (onQAClear) {
-                    onQAClear(!err, err);
-                  }
-                };
-                db.clear(qaClear)
-                  .then(() => cleared())
-                  .catch(cleared);
-              }
-              else if (qaGroup) {
-                db.leaveGroup().then(() => db.joinGroup(qaGroup));
-              }
-            }
-          })
-          .catch(ui.setError);
+        setUnitAndProblem (stores, unitCode, problemId).then( () => {
```

Comment:
UnitCodes and ProblemOrdinals can be set in two different ways:

1. via URLParam for the current window.  This happens when a *student* launches a clue activity from the portal.
2. via the `offering.activity_url` parameter when the teacher dashboard is loading activities.

To support this late-stage changing of the problem and unit (as happens in teacher dashboard), a new method `setUnitAndProblem` had to be extracted.  This changed the application initialization significantly.

While we were in here, we extracted `initRollbar` and `resolveAppMode` to simplify readability.